### PR TITLE
feat(cdc): map upstream table schema automatically for cdc table

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Automatically map the upstream table schema when creating a CDC table from `mysql-cdc` or `postgres-cdc` source.
We use the [sea-schema](https://github.com/SeaQL/sea-schema) crate to read the upstream schema.

- Add a `ExternalTableImpl` as an interface to read external table schema
- Some refactor to move mysql related code to `mysql.rs`.
- The `*` cannot be used with column definitions, for example `create table t (v1 int, *)` is prohibited.

Example:
```sql
dev=> create table supplier (*) from pg_source table 'public.supplier';
CREATE_TABLE
dev=> describe supplier;
       Name        |       Type        | Is Hidden | Description
-------------------+-------------------+-----------+-------------
 s_suppkey         | bigint            | false     |
 s_name            | character varying | false     |
 s_address         | character varying | false     |
 s_nationkey       | bigint            | false     |
 s_phone           | character varying | false     |
 s_acctbal         | numeric           | false     |
 s_comment         | character varying | false     |
 primary key       | s_suppkey         |           |
 distribution key  | s_suppkey         |           |
 table description | supplier          |           |
(10 rows)
```

close https://github.com/risingwavelabs/risingwave/issues/16443

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/source/cdc/external/mod.rs [hunk 1]
```diff
@@ -569,6 +318,40 @@ impl ExternalTableReaderImpl {
     }
 }
 
+pub enum ExternalTableImpl {
```
Comment (StrikeW) [inline src/connector/src/source/cdc/external/mod.rs:321]: Interface to connect and read external table schema

**Review by st1page — APPROVED** [review]
Generally LGTM
It is recommended to split "code movement and refactoring" and "functional logic changes" into two separate PRs next time

**Review by st1page — COMMENTED** [review]
Btw, Can you show an error message indicating that the database cannot be connected? I guess it would be very helpful for users to know that it is the frontend which need to be connected to the source database.

File: java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/connector/source/common/MySqlValidator.java [hunk 2]
```diff
@@ -195,10 +195,9 @@ private void validateTableSchema() throws SQLException {
                 var field = res.getString(1);
                 var dataType = res.getString(2);
                 var key = res.getString(3);
-                schema.put(field.toLowerCase(), dataType);
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/connector/source/common/MySqlValidator.java:198]: Yes. the `e2e_test/source/cdc_inline/auto_schema_map_mysql.slt` has defined a column named `c_Bigint`.

**Comment by StrikeW** [comment]
> Btw, Can you show an error message indicating that the database cannot be connected? I guess it would be very helpful for users to know that it is the frontend which need to be connected to the source database.

Here is an example

```
dev=> create table mysql_all_types (*) from mysql_mydb table 'mydb.mysql_mydb';
ERROR:  Failed to run the query

Caused by these errors (recent errors listed first):
  1: connector error
  2: pool timed out while waiting for an open connection
```

File: e2e_test/source/cdc_inline/auto_schema_map_mysql.slt [hunk 3]
```diff
@@ -0,0 +1,118 @@
+control substitution on
+
+# test case need to cover all data types
+system ok
+mysql --protocol=tcp -u root -e "DROP DATABASE IF EXISTS mytest; CREATE DATABASE mytest;"
```
Comment (BugenZhao) [inline e2e_test/source/cdc_inline/auto_schema_map_mysql.slt:5]: I think we can omit `--protocol=tcp` and `-u root` and directly use `mysql -e` in `.slt`.

https://github.com/risingwavelabs/risingwave/blob/d19d8f9ae55832430f452c2ed90fca44910dd37b/e2e_test/source_inline/cdc/mysql/mysql_create_drop.slt#L8-L9

File: e2e_test/source/cdc_inline/auto_schema_map_mysql.slt [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    "
+
+
+statement ok
+create source mysql_source with (
+ connector = 'mysql-cdc',
+ hostname = '${MYSQL_HOST:localhost}',
+ port = '${MYSQL_TCP_PORT:8306}',
+ username = 'root',
+ password = '${MYSQL_PWD:}',
```
Comment (BugenZhao) [inline e2e_test/source/cdc_inline/auto_schema_map_mysql.slt:44]: We can use `${RISEDEV_MYSQL_WITH_OPTIONS_COMMON}`.

File: e2e_test/source/cdc_inline/auto_schema_map_mysql.slt
Comment (BugenZhao) [inline e2e_test/source/cdc_inline/auto_schema_map_mysql.slt:1]: Better to organize it under the new folder `e2e_test/source_inline`.

**Review by BugenZhao — COMMENTED** [review]
Can we adopt the new inline source style introduced in #16449?

**Review by st1page — APPROVED** [review]

File: e2e_test/source/cdc_inline/auto_schema_map_mysql.slt
Comment (StrikeW) [reply e2e_test/source/cdc_inline/auto_schema_map_mysql.slt:1]: Will refactor in another pr.

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**